junit-test/shared/src/test/scala/org/scalajs/junit/utils/JUnitTest.scala

```diff
@@ -47,35 +55,131 @@ abstract class JUnitTest {
+        skippedEvent
+    )
+
+    def assumptionViolated(testName: String) = append(1, 0, 0)(
+        testStartedOutput(testName),
+        testAssumptionViolatedOutput(testName),
+        skippedEvent,
+        testFinishedOutput(testName)
+    )
+
+    def exception(testName: String, msg: String, clazz: String): OutputBuilder = {
+      append(1, 0, 1)(
+          testStartedOutput(testName),
+          testExceptionMsgOutput(testName, msg, clazz),
+          failureEvent,
+          testFinishedOutput(testName)
+      )
     }
 
-    def formattedAssumptionViolatedException: String =
-      "org.junit.internal." + red("AssumptionViolatedException")
+    def assertion(testName: String, message: String): OutputBuilder = {
+      append(1, 0, 1)(
+          testStartedOutput(testName),
+          testAssertionErrorMsgOutput(testName, message),
+          failureEvent,
+          testFinishedOutput(testName)
+      )
+    }
+
+    private def append(t: Int, i: Int, f: Int)(out: Output*) = {
+      new OutputBuilder(argInfo, total + t, ignored + i, failed + f, output ++ out)
+    }
+
+    // Test method.
+
+    private[JUnitTest] def checkOutput(actual: List[Output]): Unit = {
+      val expected = (
+          testRunStartedOutput :: output :::
+          List(testRunFinishedOutput(total, ignored, failed), done)
+      )
+
+      @tailrec def minimizeDiff(list1: List[Output], list2: List[Output],
+          dropped: Int): (List[Output], List[Output], Int) = {
+        (list1, list2) match {
+          case (x :: xs, y :: ys) if x == y => minimizeDiff(xs, ys, dropped + 1)
+          case _                            => (list1, list2, dropped)
+        }
+      }
+  
+      val (expected1, actual1, droppedFront) =
+        minimizeDiff(expected, actual, 0)
+      val (expected2, actual2, droppedBack) =
+        minimizeDiff(expected1.reverse, actual1.reverse, 0)
+      val expectedMinimized = expected2.reverse
+      val actualMinimized = actual2.reverse
+  
+      if (expectedMinimized != actualMinimized) {
+        val msg = new StringBuilder
+        def appendElems(original: List[Output], minimized: List[Output]): Unit = {
+          def appendGreyLine(out: Output): Unit = {
+            msg.append(GREY + "  " + withoutColor(out.toString) + NORMAL + "\n")
+          }
+          original.slice(droppedFront - 3, droppedFront).foreach(appendGreyLine)
+          minimized.foreach(out => msg.append("  " + out + "\n"))
+          val backIndex = original.size - droppedBack
+          original.slice(backIndex, backIndex + 3).foreach(appendGreyLine)
+        }
+  
+        msg.append(s"JUnit output mismatch with $argInfo:\n")
+        msg.append(s"Expected: List(\n")
+        appendElems(expected, expectedMinimized)
+        msg.append(")\nbut got: List(\n")
+        appendElems(actual, actualMinimized)
+        msg.append(")")
+  
+        throw new Exception(msg.result())
+      }
+    }
+
+    // Text builders.
 
     def done: Output = Done("")
```

Comment:
Looks like the methods returning `Output`s could become private, couldn't they? Maybe even `Output` itself can become private?